### PR TITLE
Move Dense and Dropout layers to separate files

### PR DESCRIPTION
**Pull request by rickwierenga**
For details see: https://groups.google.com/a/tensorflow.org/forum/m/#!topic/swift/-I-P8wjhsTg

### CODE REVIEW

**Comment by rickwierenga** [comment]
@googlebot I signed it!

**Comment by rickwierenga** [comment]
I just noticed (too late) that this PR also includes the precondition described [here](https://groups.google.com/a/tensorflow.org/forum/m/#!topic/swift/Hi42LFqe_s4). Is this a problem?

**Comment by Shashi456** [comment]
@rickwierenga, it's ideal if we keep the splitting of core.swift into separate files and this precondition as separate PRs

File: Sources/TensorFlow/Layers/Core.swift [hunk 1]
```diff
@@ -12,67 +12,19 @@
...
-    @noDerivative public let probability: Double
-
-    /// Creates a dropout layer.
-    ///
-    /// - Parameter probability: The drop probability.
-    public init(probability: Double) {
-        self.probability = probability
-    }
+/// A layer that encloses a custom differentiable function.
+public struct Function<Input: Differentiable, Output: Differentiable>: ParameterlessLayer {
```
Comment (saeta) [inline Sources/TensorFlow/Layers/Core.swift]: Could you keep the ordering of the layers the same as before? (This will help make the history easier to follow.)

File: Sources/TensorFlow/Layers/Dropout.swift [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+public struct Dropout<Scalar: TensorFlowFloatingPoint>: ParameterlessLayer {
+    @noDerivative public let probability: Double
+
+    /// Creates a dropout layer.
+    ///
+    /// - Parameter probability: The drop probability.
+    /// - Precondition: probability must be a value between 0 and 1 (inclusive). 
+    public init(probability: Double) {
+        precondition(0...1 ~= probability,
+            "Probability must be a value between 0 and 1 (inclusive) but is \(probability)")
```
Comment (saeta) [inline Sources/TensorFlow/Layers/Dropout.swift]: FWIW, although it's generally better to have one logical change per PR, given that this is just one statement, I think it's probably okay to leave in. 

Note: this change would break users who relied upon probability outside the range acting as if it's clamped to the range. I think this is okay, however, as this likely indicates a bug. That said, if we wanted to, we could provide a convenience initializer that would do the clamping to 0...1 and then delegate to this initializer. (This could be done in a follow-on PR.)

Comment (rickwierenga) [reply Sources/TensorFlow/Layers/Dropout.swift]: > FWIW, although it's generally better to have one logical change per PR, given that this is just one statement, I think it's probably okay to leave in.

You are absolutely right. I forgot to exclude it in the commit--will look more closely in the future.

> Note: this change would break users who relied upon probability outside the range acting as if it's clamped to the range. I think this is okay, however, as this likely indicates a bug. That said, if we wanted to, we could provide a convenience initializer that would do the clamping to 0...1 and then delegate to this initializer. (This could be done in a follow-on PR.)

The convenience initializer ia a great idea. I think it should show some kind of message though indicating the input value will be clamped. Should this be a permanent or temporary initializer (with deprecation warning)?

**Comment by rickwierenga** [comment]
Just fixed the ordering issue. I will leave the changes in Dropout.swift for the next PR.

File: Sources/TensorFlow/Layers/Dropout.swift [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    public func callAsFunction(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        switch Context.local.learningPhase {
+        case .training:
+            return applyingTraining(to: input)
+        case .inference:
+            return applyingInference(to: input)
+        }
+    }
+}
+
```
Comment (dan-zheng) [inline Sources/TensorFlow/Layers/Dropout.swift]: Please remove empty lines.

File: Sources/TensorFlow/Layers/Dense.swift [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        weightInitializer: ParameterInitializer<Scalar> = glorotUniform(),
+        biasInitializer: ParameterInitializer<Scalar> = zeros()
+    ) {
+        self.init(
+            weight: weightInitializer([inputSize, outputSize]),
+            bias: biasInitializer([outputSize]),
+            activation: activation)
+    }
+}
+
```
Comment (dan-zheng) [inline Sources/TensorFlow/Layers/Dense.swift]: Please remove empty lines here, too.
